### PR TITLE
prusti_assert!

### PR DESCRIPTION
**Pull request by JM4ier**
a macro to allow full prusti syntax in assertions (WIP)

### CODE REVIEW

File: prusti-interface/src/specs/mod.rs [hunk 1]
```diff
@@ -47,6 +47,7 @@ pub struct SpecCollector<'a, 'tcx: 'a> {
     /// Map from functions/loops and their specifications.
     procedure_specs: HashMap<LocalDefId, ProcedureSpecRefs>,
     loop_specs: Vec<LocalDefId>,
+    prusti_assertion_specs: Vec<LocalDefId>,
```
Comment (Aurel300) [inline prusti-interface/src/specs/mod.rs]: Why `_specs`?

File: prusti-interface/src/specs/typed.rs [hunk 2]
```diff
@@ -83,6 +88,11 @@ pub struct LoopSpecification {
     pub invariant: LocalDefId,
 }
 
+#[derive(Debug, Clone)]
+pub struct PrustiAssertionSpecification {
```
Comment (Aurel300) [inline prusti-interface/src/specs/typed.rs]: Here also: why `Specification`?

File: prusti-tests/tests/verify_overflow/fail/prusti_assert.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    body_invariant!(false);
+}
+
+fn loop_shouldnt_crash() {
+    loop {
+        prusti_assert!(true);
+        body_invariant!(true);
+    }
+}
+
```
Comment (Aurel300) [inline prusti-tests/tests/verify_overflow/fail/prusti_assert.rs:51]: Add tests with quantifiers and implications.

File: prusti-viper/src/encoder/mir/specifications/specs.rs [hunk 4]
```diff
@@ -74,6 +75,11 @@ impl<'tcx> Specifications<'tcx> {
         self.user_typed_specs.get_loop_spec(def_id)
     }
 
+    pub(super) fn get_assert_spec(&self, def_id: &DefId) -> Option<&PrustiAssertionSpecification> {
+        trace!("Get loop specs of {:?}", def_id);
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir/specifications/specs.rs]: Change comment.

**Comment by Aurel300** [comment]
Looks good so far. For the desugaring to actually do something, you should probably pass the closure created in the rewriter into some special method added to `prusti-contracts` (similar to e.g. `forall`), then intercept the call to it in the procedure encoder and the pure function encoder.

Please also add `prusti_assume!`, which should be basically the same thing.

File: prusti-viper/src/encoder/errors/error_manager.rs [hunk 5]
```diff
@@ -146,6 +146,8 @@ pub enum ErrorCtxt {
     CloseFracRef,
     /// Failed to set an active variant of an union.
     SetEnumVariant,
+    /// Failed Assertion
+    Assertion,
```
Comment (Aurel300) [inline prusti-viper/src/encoder/errors/error_manager.rs]: Would this not be better as another case in `PanicCause`?

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 6]
```diff
@@ -1781,13 +1781,46 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+        &mut self,
+        bb: mir::BasicBlock,
+        block: &mir::BasicBlockData<'tcx>,
+        encoded_statements: &mut Vec<vir_high::Statement>,
+    ) -> SpannedEncodingResult<bool> {
+        for stmt in &block.statements {
+            if let mir::StatementKind::Assign(box (
+                _,
+                mir::Rvalue::Aggregate(box mir::AggregateKind::Closure(cl_def_id, cl_substs), _),
+            )) = stmt.kind
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:1841]: Can you tell these apart from loop invariants? Maybe if the `get_prusti_assertion_specs` call returned `None`, this should return `Ok(false)`?

File: prusti-tests/tests/verify_overflow/fail/prusti_assert.rs (hunk 3 above)
Comment (vakaras) [reply prusti-tests/tests/verify_overflow/fail/prusti_assert.rs:51]: Since lowering of quantifiers is not yet supported, let's land it as it is.

**Comment by Aurel300** [comment]
@vakaras This PR should also handle the old encoding at some point, right?

**Comment by vakaras** [comment]
> @vakaras This PR should also handle the old encoding at some point, right?

No. I do not think Jonas has enough time for that.

**Review by vakaras — APPROVED** [review]
LGTM. Thanks!

bors r+

**Pull request merged**